core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -237,13 +238,23 @@ class ProducerIdMapping(val topicPartition: TopicPartition,
    * Update the mapping to the given epoch and sequence number.
    */
   def update(pid: Long, pidEntry: PidEntry): Unit = {
-    if (pid != LogEntry.NO_PID) {
+    if (pid >= LogEntry.NO_PID) {
```

Comment:
This is actually wrong. It should be `pid > LogEntry.NO_PID`. Otherwise `-1` is an accepted PID, which is not what we discussed. 

I think we should have a public static method like `LogEntry.isPidValid(long pid)` so we don't have different notions of valid in different parts of the code base.